File: android/app/src/main/kotlin/com/cerashealth/ceras/lifeplus/B369Device.kt
package com.cerashealth.ceras.lifeplus

import android.content.Context
import android.util.Log
import cn.icomon.icdevicemanager.ICDeviceManager
import cn.icomon.icdevicemanager.ICDeviceManagerDelegate
import cn.icomon.icdevicemanager.ICDeviceManagerSettingManager
import cn.icomon.icdevicemanager.model.data.*
import cn.icomon.icdevicemanager.model.device.ICDevice
import cn.icomon.icdevicemanager.model.device.ICDeviceInfo
import cn.icomon.icdevicemanager.model.device.ICScanDeviceInfo
import cn.icomon.icdevicemanager.model.device.ICUserInfo
import cn.icomon.icdevicemanager.model.other.ICConstant.*
import cn.icomon.icdevicemanager.model.other.ICDeviceManagerConfig
import com.cerashealth.ceras.MainActivity
import com.cerashealth.ceras.lifeplus.data.ConnectionInfo
import com.cerashealth.ceras.lifeplus.data.WeightData
import com.walnutin.hardsdk.ProductList.sdk.HardSdk
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.util.*


class B369Device :BaseDevice(), ICDeviceManagerDelegate{

    companion object{
        val TAG = B369Device::class.java.simpleName
        var device:ICDevice? = null
        var b369Device:B369Device? = null
        var deviceConnected = false
        var initComplete = false
        var deviceFound = false
        var deviceId:String? = null
        var result:MethodChannel.Result? = null

        fun getInstance():B369Device{
            if(b369Device==null){
                b369Device = B369Device()
            }
            return b369Device!!
        }
    }

    fun initSDK(context: Context?)
    {
        val config = ICDeviceManagerConfig()
        config.context = context
        val userInfo = ICUserInfo()
        userInfo.kitchenUnit = ICKitchenScaleUnit.ICKitchenScaleUnitOz
        userInfo.rulerUnit = ICRulerUnit.ICRulerUnitInch
        userInfo.weightUnit = ICWeightUnit.ICWeightUnitLb
        userInfo.age = 31
        userInfo.weight = 180.0
        userInfo.height = 170
        userInfo.userIndex = 1
        userInfo.sex = ICSexType.ICSexTypeMale
        userInfo.weightDirection = 1
        userInfo.targetWeight = 160.0

        config.is_fill_adc = true
        ICDeviceManager.shared().delegate = this
        ICDeviceManager.shared().setUserList(listOf(userInfo))
        ICDeviceManager.shared().updateUserInfo(userInfo)
        ICDeviceManager.shared().initMgrWithConfig(config)
    }



    override fun connectDevice(context: Context, result: MethodChannel.Result, deviceId: String?) {
        Log.i(TAG, "Connecting to $deviceId")
        if(!initComplete) {
            initSDK(MainActivity.currentActivity)
        }
//        Log.d(TAG,"Bluetooth enabled ${ICDeviceManager.shared().isBLEEnable}")

//        if(ICDeviceManager.shared().isBLEEnable){
            Log.d(TAG,"Bluetooth enabled")
            Log.d(TAG,"Setting call back")

            GlobalScope.launch {
                if(!initComplete) {
                    //If the device is initialized then wait for 5 secs
                    //Scan the device and update
                    delay(5000)
                }
                ICDeviceManager.shared().scanDevice {
                    onScanResult(it, result, deviceId)
                }

//            }

            //This is used to wait for 30 seconds and send error back
            GlobalScope.launch {
                delay(30000)
                ICDeviceManager.shared().stopScan()
                Log.d(TAG, "Device connected $deviceConnected")
                //If the device is not connected, then send an error back
                if(!deviceConnected) {
                    Log.d(TAG, "Device found $deviceFound and Device connected $deviceConnected")
                    MainActivity.currentActivity?.runOnUiThread {
                        result.success(ConnectionInfo.createResponse(message = "Failed", connected = false, deviceFound = deviceFound))
                    }
                }
            }
        }
    }

    override fun connectWifi(result: MethodChannel.Result?, connectionInfo: ConnectionInfo, context: Context, network: String?, password: String?) {
        Log.d(TAG,"Connecting to $network with $password with device id ${device?.macAddr}")
        B369Device.result = result
        ICDeviceManager.shared().settingManager.configWifi(device, network!!, password!!) {
            Log.d(TAG, "Got wifi result status $it")
            when (it) {
                ICSettingCallBackCode.ICSettingCallBackCodeSuccess -> {
                    Log.d(TAG, "Connection success")
                    DataSync.getServerUrl()?.let {
                        updateServerUrl(it)
                    }
                }
                else -> {
                    Log.d(TAG, "Connection failure")
                    result?.success(
                        ConnectionInfo.createResponse(
                            message = "Failure ${it}",
                            connected = true
                        )
                    )
                }
            }
        }

        //Wait for 30 seconds and send an error response back if it is not connected
        GlobalScope.launch {
            delay(40000)
            try {
                B369Device.result?.let {
                    MainActivity.currentActivity?.runOnUiThread {
                        sendWifiResponse(message = "Config failed", connected = false)
                    }
                }
            }catch (e:Exception){
                Log.e(TAG,"Error while sending connect")
            }
        }

    }

    private fun updateServerUrl(serverUrl:String){
        ICDeviceManager.shared().settingManager.setServerUrl(device,serverUrl){
            Log.d(TAG, "server Url updated to $serverUrl")
        }
    }

    override fun syncData(result: MethodChannel.Result?, connectionInfo: ConnectionInfo, context: Context) {
        result?.success("Load complete")
    }

    override fun onInitFinish(finished: Boolean) {
        Log.d(TAG, "Initialization finished with status $finished")
        initComplete = finished
    }

    override fun onBleState(state: ICBleState?) {
        Log.d(TAG, "BLE state changed $state")
//        if (state == ICBleState.ICBleStatePoweredOn) {
//
//        }
    }

    override fun disconnectDevice(result: MethodChannel.Result?,deviceId:String?) {
        ICDeviceManager.shared().removeDevice(ICDevice().apply {
            macAddr = deviceId
        }) { _, icAddDeviceCallBackCode ->
            if (icAddDeviceCallBackCode == ICRemoveDeviceCallBackCode.ICRemoveDeviceCallBackCodeSuccess)
                result?.success("Success")
            else
                result?.success("Error")
        }
    }

    override fun onDeviceConnectionChanged(device: ICDevice?, connState: ICDeviceConnectState?) {
        deviceConnected = (ICDeviceConnectState.ICDeviceConnectStateConnected == connState)
        Log.d(TAG, "Connection changed $connState")
    }

    override fun onReceiveWeightData(device: ICDevice?, weightData: ICWeightData?) {
        Log.d(TAG, "received Weight data ")
        weightData?.let {
            val measureTime = Calendar.getInstance().apply {
                timeInMillis = weightData.time*1000
            }
            DataSync.uploadWeightData(listOf(WeightData(measureTime.time,weightData.weight_lb,weightData.weight_kg,device!!.macAddr)))
            Log.d(TAG, "Got weight data at ${weightData.time} weight ${weightData.weight_lb} BMI ${weightData.bmi} " +
                    "bodyfat ${weightData.bodyFatPercent} " +
                    "Musc Percent ${weightData.musclePercent} " +
                    "Viscural Fat ${weightData.visceralFat} " +
                    "Bone Bass ${weightData.boneMass} " +
                    "Body Age ${weightData.physicalAge}" +
                    "Protien ${weightData.proteinPercent}")
        }
    }

    override fun onReceiveKitchenScaleData(device: ICDevice?, p1: ICKitchenScaleData?) {
        Log.d(TAG, "Scale data updated ")
    }

    override fun onReceiveKitchenScaleUnitChanged(device: ICDevice?, p1: ICKitchenScaleUnit?) {
        Log.d(TAG, "Scale unit changed ")
    }

    override fun onReceiveCoordData(device: ICDevice?, p1: ICCoordData?) {
        Log.d(TAG, "Got Cord Data ")
    }

    override fun onReceiveRulerData(device: ICDevice?, p1: ICRulerData?) {
        Log.d(TAG, "Got ruler data ")
    }

    override fun onReceiveRulerHistoryData(device: ICDevice?, p1: ICRulerData?) {
        Log.d(TAG, "Got ruler history ")
    }

    override fun onReceiveWeightCenterData(device: ICDevice?, p1: ICWeightCenterData?) {
        Log.d(TAG, "GOt weigh center data")
    }

    override fun onReceiveWeightUnitChanged(device: ICDevice?, p1: ICWeightUnit?) {
        Log.d(TAG, "Weight unit changed ")
    }

    override fun onReceiveRulerUnitChanged(device: ICDevice?, p1: ICRulerUnit?) {
        Log.d(TAG, "Ruler unit changed ")
    }

    override fun onReceiveRulerMeasureModeChanged(device: ICDevice?, p1: ICRulerMeasureMode?) {
        Log.d(TAG, "Ruler mode changed ")
    }

    override fun onReceiveMeasureStepData(device: ICDevice?, step: ICMeasureStep?, data2: Any?) {
        when (step) {
            ICMeasureStep.ICMeasureStepMeasureWeightData -> {
                val data = data2 as ICWeightData
            }
            ICMeasureStep.ICMeasureStepMeasureCenterData -> {
                val data = data2 as ICWeightCenterData
                onReceiveWeightCenterData(device, data)
            }
            ICMeasureStep.ICMeasureStepAdcStart -> {
                Log.d(TAG, device!!.getMacAddr() + ": start imp... ")
            }
            ICMeasureStep.ICMeasureStepAdcResult -> {
                Log.d(TAG, device!!.getMacAddr() + ": imp over")
            }
            ICMeasureStep.ICMeasureStepHrStart -> {
                Log.d(TAG, device!!.getMacAddr() + ": start hr")
            }
            ICMeasureStep.ICMeasureStepHrResult -> {
                val hrData = data2 as ICWeightData
                Log.d(TAG, device!!.getMacAddr() + ": over hr: " + hrData.hr)
            }
            ICMeasureStep.ICMeasureStepMeasureOver -> {
                val data = data2 as ICWeightData
                Log.d(TAG, device!!.getMacAddr() + ": over measure")
                onReceiveWeightData(device, data)
            }
            else -> {
            }
        }

    }

    override fun onReceiveWeightHistoryData(device: ICDevice?, historyData: ICWeightHistoryData?) {
        historyData?.let {historyData->
            Log.d(TAG, "Got history data ${historyData.time} ${historyData.weight_lb} ")
            val measureTime = Calendar.getInstance().apply {
                timeInMillis = historyData.time*1000L
            }
            DataSync.uploadWeightData(listOf(WeightData(measureTime.time,historyData.weight_lb,historyData.weight_kg,device!!.macAddr)))
        }
        Log.d(TAG, "Weight history data ")
    }

    override fun onReceiveSkipData(device: ICDevice?, p1: ICSkipData?) {
        Log.d(TAG, "Skip device ")
    }

    override fun onReceiveHistorySkipData(device: ICDevice?, p1: ICSkipData?) {
        Log.d(TAG, "History skip data ")
    }

    override fun onReceiveSkipBattery(device: ICDevice?, p1: Int) {
        Log.d(TAG, "Skip dattery ")
    }

    override fun onReceiveUpgradePercent(device: ICDevice?, p1: ICUpgradeStatus?, p2: Int) {
        Log.d(TAG, "Device upgrade percent ")
    }

    override fun getDeviceInfo(result: MethodChannel.Result?, connectionInfo: ConnectionInfo, context: Context) {
        sendConnectionResponse(connectionInfo.deviceId, deviceConnected, result)
    }

    override fun onReceiveDeviceInfo(device: ICDevice?, deviceInfo: ICDeviceInfo?) {
        Log.d(TAG, "Got device info")
    }

    override fun onReceiveDebugData(device: ICDevice?, p1: Int, p2: Any?) {
        Log.d(TAG, "Got debug data ")
    }

    private fun sendWifiResponse(message:String, connected:Boolean){
        try {
            result?.success(ConnectionInfo.createResponse(message = message, connected = connected))
            result = null
        }catch (e:Exception){
            Log.e(TAG,"Error while sending response ",e)
        }
    }

    override fun onReceiveConfigWifiResult(device: ICDevice?, wifiStatus: ICConfigWifiState?) {
        Log.i(TAG,"Got wifi result for ${device?.macAddr} is $wifiStatus")
        result?.let {
            Log.i(TAG,"Result is not null sending the response back")
            when(wifiStatus){
                ICConfigWifiState.ICConfigWifiStateSuccess ->{
                    sendWifiResponse(message = "Wifi Connected",connected = true)
                }
                ICConfigWifiState.ICConfigWifiStateWifiConnectFail ->{
                    Log.i(TAG,"Sending wifi fail ")
                    sendWifiResponse(message = "Config failed",connected = false)
                }
                ICConfigWifiState.ICConfigWifiStateFail -> {
                    Log.i(TAG,"Sending wifi fail ")
                    sendWifiResponse(message = "Config failed",connected = false)
                }
                ICConfigWifiState.ICConfigWifiStatePasswordFail -> {
                    Log.i(TAG,"Sending wifi password fail ")
                    sendWifiResponse(message = "Invalid Password",connected = false)
                }
                ICConfigWifiState.ICConfigWifiStateWifiConnecting -> {
                    Log.d(TAG,"Connecting ${device?.macAddr}")
                }
                else->{
                    Log.d(TAG,"Got state $wifiStatus")
                }
            }
        }
    }

    override fun getConnectionStatus(result: MethodChannel.Result?, connectionInfo: ConnectionInfo, context: Context){
        result?.success(deviceConnected)
    }

    /**
     * This method takes the scan result and validates if it matches the deviceId
     */
    private fun onScanResult(deviceInfo: ICScanDeviceInfo?,scanResult: MethodChannel.Result?,requestedDeviceId:String?) {
        Log.i(TAG, "Found device ${deviceInfo?.macAddr}")
        deviceInfo?.let {
            //Capture the last 5 charecters of the mac address and remove the ":" to get the last four
            val lastFour = it.macAddr.substring(it.macAddr.length - 5).replace(":", "")

            Log.d(TAG,"matching $lastFour to $requestedDeviceId")
            //If the last Four characters matches the requested device id
            if(lastFour.equals(requestedDeviceId, ignoreCase = true)){
                Log.d(TAG, "values matched matched")
                deviceFound = true
                scanResult?.let {
                    val device = ICDevice().apply {
                        macAddr = deviceInfo.macAddr

                    }
                    //Device is found, stop the scan
                    ICDeviceManager.shared().stopScan()
                    //Add the device to the devices
                    ICDeviceManager.shared().addDevice(device) { device, _ ->
                        try {
                            B369Device.device = device
                            scanResult.let {
                                scanResult.success(ConnectionInfo.createResponse(message = "Connected", connected = true, deviceId = deviceInfo.macAddr, deviceName = deviceInfo.name,
                                    deviceType = B369_DEVICE, deviceFound = deviceFound))
                            }

                            deviceConnected = true
                            Log.d(TAG,"Setup the device with device connected $deviceConnected")

                            //Set the scale to pounds, this has to be configurable based on the user preference
//                            ICDeviceManager.shared().settingManager.setScaleUnit(device, ICWeightUnit.ICWeightUnitLb) { code ->
//                                Log.d(TAG, String.format("%s %s", device.getMacAddr(),"setting callback code :$code"))
//                            }
                            Log.d(TAG,"Log path ${ICDeviceManager.shared().logPath}")

                        } catch (ex: Exception) {
                            //If there is an exception, let the timeout catch this.
                            Log.e(TAG, "Error while sending response ", ex)
                        }
                    }
                }
            }
        }
    }

}

File: ios/Runner/B369Device.swift
//
//  B369Device.swift
//  Runner
//
//  Created by Shyam Alapati on 7/15/21.
//

import Foundation
import ICDeviceManager

class B369Device: NSObject,ICScanDeviceDelegate,ICDeviceManagerDelegate{
    
    var device:ICDevice? = nil
    var deviceId:String? = nil
    var connectionResult:FlutterResult? = nil
    var wifiResult:FlutterResult? = nil
    var deviceFound = false
    var deviceConnected = false
    var isDeviceConnectionAvailable = false
    var initComplete = false
    private var deviceAdded = false
    
    //Initialize the B369
    override init() {
        super.init()
        let userInfo = ICUserInfo()
        userInfo.age = 30
        userInfo.enableMeasureHr = true
        userInfo.height = 170
        userInfo.sex = ICSexType.male
        userInfo.weightUnit = ICWeightUnit.lb
        userInfo.weight = 60
        userInfo.peopleType = ICPeopleTypeNormal
        userInfo.kitchenUnit = ICKitchenScaleUnit.lb
        userInfo.rulerUnit = ICRulerUnit.inch
        userInfo.userIndex = 1
        let userData = DataSync.getUserInfo()
        if(userData != nil){
            userInfo.age = UInt(userData!.age)
            userInfo.sex = userData?.sex.lowercased() == "male" ? ICSexType.male : ICSexType.femal
            userInfo.weight = Float(userData!.weightInKgs) * 2.20
            userInfo.height = UInt(userData!.heightInCm)
        }
        ICDeviceManager.shared()?.update(userInfo)
        ICDeviceManager.shared()?.delegate = self
        ICDeviceManager.shared()?.initMgr()
        deviceAdded = false
    }
    
    func startScan(result:@escaping FlutterResult,scanDeviceId:String) {
        self.deviceId = scanDeviceId
        self.deviceFound=false
        self.deviceConnected=false
        self.connectionResult = result
        if(initComplete){
            self.scanAndFindDevice(result: result)
        }else{
            DispatchQueue.main.asyncAfter(deadline: .now() + 5) {
                self.scanAndFindDevice(result: result)
            }
        }
    }
    
    private func scanAndFindDevice(result:@escaping FlutterResult){
        ICDeviceManager.shared()?.scanDevice(self)
        //Stop scanning after 20 seconds
        DispatchQueue.main.asyncAfter(deadline: .now() + 20) {
            self.sendResponse(result: result)
        }
    }
    
    func syncData(connectionInfo:ConnectionInfo,result:@escaping FlutterResult){
        self.checkAndReAddDevice(macAddr: connectionInfo.deviceId)
        result(generateResponse(deviceId: connectionInfo.deviceId!, message: "Synced", connected: isDeviceConnectionAvailable))
    }
    
    func connectWifi(result:@escaping FlutterResult,ssid:String,password:String){
        NSLog("Conecting to \(ssid) with password \(password)")
        self.wifiResult = result
        ICDeviceManager.shared()?.getSettingManager()?.configWifi(self.device, ssid: ssid, password: password, callback: { (callBackCode) in
            NSLog("Got connection response \(callBackCode.rawValue)")
        })
        
        let serverBaseUrl = UserDefaults.standard.object(forKey: AppDelegate.SERVER_BASE_URL) as! String
        NSLog("Sending server base url \(String(describing: serverBaseUrl))")
        ICDeviceManager.shared()?.getSettingManager()?.setServerUrl(self.device, server: serverBaseUrl, callback: { (callBackCode) in
            NSLog("Got Server URL \(callBackCode.rawValue)")
        })
        DispatchQueue.main.asyncAfter(deadline: .now() + 45) {
            self.sendWifiResponse(device: self.device,message: "Config failed",connected: false)
        }
    }
    
    func onReceiveDeviceInfo(_ device: ICDevice!, deviceInfo: ICDeviceInfo!) {
        NSLog("Got device info firmware :\(String(deviceInfo.firmwareVer)) serial #\(String(deviceInfo.sn))")
    }
    
    func disconnect(result:@escaping FlutterResult,connectionInfo: ConnectionInfo){
        let deviceForDeletion = ICDevice()
        deviceForDeletion.macAddr = connectionInfo.deviceId
        ICDeviceManager.shared()?.remove(deviceForDeletion, callback: { device, callBack in
            if(callBack == ICRemoveDeviceCallBackCode.success){
                result("Success")
            }else{
                result("Error")
            }
        })
    }
    
    func onReceiveConfigWifiResult(_ device: ICDevice!, state: ICConfigWifiState) {
        NSLog("Got Wifi config result \(state.rawValue) for device \(String(device.macAddr))")
        switch state {
        case ICConfigWifiState.wifiConnectFail:
            sendWifiResponse(device: device,message: "Config failed",connected: false)
        case ICConfigWifiState.passwordFail:
            sendWifiResponse(device: device,message: "Invalid password",connected: false)
        case ICConfigWifiState.success:
            sendWifiResponse(device: device,message: "success",connected: true)
        case ICConfigWifiState.fail:
            sendWifiResponse(device: device,message: "Config failed",connected: false)
        case ICConfigWifiState.wifiConnecting:
            NSLog("Connecting to wifi")
        case ICConfigWifiState.serverConnecting:
            NSLog("Connecting to device")
        default:
            sendWifiResponse(device: device,message: "Config failed",connected: false)
        }
    }
    
    func sendWifiResponse(device: ICDevice!,message: String,connected: Bool){
        if(self.wifiResult != nil){
            self.wifiResult?(generateResponse(deviceId: device.macAddr, message:message,connected: connected))
            self.wifiResult = nil
        }
    }
    
    
    func onBleState(_ state: ICBleState) {
        NSLog("Got Ble state \(state)")
    }
    
    func onDeviceConnectionChanged(_ device: ICDevice!, state: ICDeviceConnectState) {
        if(state == ICDeviceConnectState.connected){
            NSLog("Got connected changed to true")
            isDeviceConnectionAvailable = true
        }else if(state == ICDeviceConnectState.disconnected){
            NSLog("Got connected changed to false")
            isDeviceConnectionAvailable = false
        }
        NSLog("Got connection changed to \(state)")
    }
    
    func onInitFinish(_ bSuccess: Bool) {
        NSLog("Init finish status \(bSuccess)")
        if(bSuccess){
            self.initComplete = true
            if(self.deviceId != nil){
                checkAndReAddDevice(macAddr: self.deviceId)
            }
        }
    }
    
    func getConnectionStatus(result:@escaping FlutterResult,connectionInfo: ConnectionInfo){
        self.checkAndReAddDevice(macAddr: connectionInfo.deviceId)
        return result(isDeviceConnectionAvailable)
    }
    
    private func checkAndReAddDevice(macAddr:String?){
        if(initComplete == true && deviceAdded == false){
            let connectedDevice = ICDevice()
            connectedDevice.macAddr = macAddr
            NSLog("Device found and adding to devices")
            ICDeviceManager.shared()?.add(connectedDevice, callback: { (device, callbackCode) in
                NSLog("Got call back code \(callbackCode.rawValue) for \(device?.macAddr ?? "")")
                if(callbackCode == .success){
                    self.deviceAdded = true
                }
            })
        }else if(deviceId == nil){
            self.deviceId = macAddr
        }
    }
    
    func getCurrentDeviceStatus(connInfo: ConnectionInfo, result:@escaping FlutterResult){
        do{
            self.checkAndReAddDevice(macAddr: connInfo.deviceId)
            let connectionInfo = ConnectionInfo(deviceId: connInfo.deviceId, deviceName: connInfo.deviceName, connected: isDeviceConnectionAvailable, deviceFound: connInfo.deviceFound, message: "",batteryStatus: "99")
            let deviceJson = try JSONEncoder().encode(connectionInfo)
            let connectionInfoData = String(data: deviceJson, encoding: .utf8)!
            NSLog("Sending Connection data back from device info \(connectionInfoData)")
            result(connectionInfoData)
        }catch{
            NSLog("Error reading data \(error)")
            result("Error")
        }
    }
    
    private func generateResponse(deviceId:String,message:String,connected: Bool) -> String{
        let connectionInfo = ConnectionInfo(deviceId: deviceId, deviceName: "", connected: connected, deviceFound: self.deviceFound, message: message)
        do{
            let deviceJson = try JSONEncoder().encode(connectionInfo)
            let connectionInfoData = String(data: deviceJson, encoding: .utf8)!
            return connectionInfoData
        }catch{
            return "Error"
        }
    }
    
    func onReceiveWeightUnitChanged(_ device: ICDevice!, unit: ICWeightUnit) {
        NSLog("Weight unit changes to \(unit)")
    }
    
    func onReceiveWeightData(_ device: ICDevice!, data: ICWeightData!) {
        NSLog("Got Weight data \(data)")
        updateWeight(device: device, weight: data)
    }
    
    func updateWeight(device: ICDevice!,weight: ICWeightData){
        DataSync.uploadWeights(weights:[WeightUpload(measureTime: Date(), kgs: weight.weight_kg, lbs: weight.weight_lb, deviceId: device.macAddr)])
    }
    
    
    
    func onReceiveWeightHistoryData(_ device: ICDevice!, data: ICWeightHistoryData!) {
        NSLog("Got weight history data \(data)")
    }
    
    func onReceiveCoordData(_ device: ICDevice!, data: ICCoordData!) {
        NSLog("Got Corod data \(data)")
    }
    
    func onReceiveMeasureStepData(_ device: ICDevice!, step: ICMeasureStep, data: NSObject!) {
        NSLog("Got measure data \(step) \(data)")
        switch step {
        case ICMeasureStepMeasureOver :
            NSLog("got measure over data \(data)")
            let weightData = data as! ICWeightData
            self.updateWeight(device: device, weight: weightData)
        case ICMeasureStepMeasureWeightData:
            NSLog("got center weight data \(data)")
        case ICMeasureStepHrStart:
            NSLog("For Step Hr start \(data)")
        case ICMeasureStepAdcResult:
            NSLog("Got adc result \(data)")
        case ICMeasureStepAdcStart:
            NSLog("Adc start \(data)")
        case ICMeasureStepHrResult:
            NSLog("Hr result \(data)")
        default:
            NSLog("Default case")
        }
    }
    
    func sendResponse(result:@escaping FlutterResult){
        ICDeviceManager.shared().stopScan()
        if(!self.deviceConnected){
            var connectionInfo = ConnectionInfo(deviceId: "", deviceName: "", connected: false, deviceFound: self.deviceFound, message: "failed")
            do{
                let deviceJson = try JSONEncoder().encode(connectionInfo)
                let connectionInfoData = String(data: deviceJson, encoding: .utf8)!
                NSLog("Sending Connection data back from device info \(connectionInfoData)")
                result(connectionInfoData)
            }catch{
                NSLog("Error getting watch info from getDeviceInfo \(error)")
                result("Error")
            }
        }
    }

    
    func onScanResult(_ deviceInfo: ICScanDeviceInfo!) {
        self.deviceFound = true
        NSLog("Got mac address \(deviceInfo.macAddr) and matching it with \(deviceId)")
        let macAddress = deviceInfo.macAddr.suffix(5).replacingOccurrences(of: ":", with: "")
        NSLog("mathcing the updated mac address to  \(macAddress) and matching it with \(deviceId)")
        if(macAddress == deviceId){
            //Once the device is found stop the scan
            ICDeviceManager.shared().stopScan()
            let connectedDevice = ICDevice()
            connectedDevice.macAddr = deviceInfo.macAddr
            NSLog("Device found and adding to devices")
            ICDeviceManager.shared()?.add(connectedDevice, callback: { (device, callbackCode) in
                NSLog("Got call back code \(callbackCode) for \(device?.macAddr)")
                self.deviceConnected = true
                self.device = device
                let connectionInfo = ConnectionInfo(deviceId: device?.macAddr, deviceName: deviceInfo.name, connected: true, deviceFound: self.deviceFound, message: "connected",deviceType: AppDelegate.SCALE_TYPE)
                do{
                    let deviceJson = try JSONEncoder().encode(connectionInfo)
                    let connectionInfoData = String(data: deviceJson, encoding: .utf8)!
                    NSLog("Sending connection info \(connectionInfoData)")
                    self.connectionResult?(connectionInfoData)
                }catch{
                    NSLog("Got error while adding device \(error)")
                    self.connectionResult?("Error")
                }
            })
        }
    }

}
